File: sync_classifications.js
const fs = require('fs');
const https = require('https');

const csvFile = 'PP_ KHSX TOTAL 26 - DATA Đặt Hàng 2026.csv';
const url = 'https://script.google.com/macros/s/AKfycbyCUwPQbji4QRXS4E7KRQ3PERxlu-IByYSdJCXgeLucxFupJukqLq_0CdXKZpC7okKHsQ/exec';

const content = fs.readFileSync(csvFile, 'utf8');

/**
 * Properly parse CSV with multi-line quoted fields.
 * The old approach of content.split('\n') breaks when CSV fields
 * contain newlines inside quotes (e.g., "8/12/2025 ( RV)\nRV 11/3/26").
 */
function parseCSV(text) {
  const rows = [];
  let currentRow = [];
  let currentVal = '';
  let inQuote = false;
  
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    
    if (inQuote) {
      if (c === '"') {
        if (i + 1 < text.length && text[i + 1] === '"') {
          currentVal += '"';
          i++;
        } else {
          inQuote = false;
        }
      } else {
        currentVal += c;
      }
    } else {
      if (c === '"') {
        inQuote = true;
      } else if (c === ',') {
        currentRow.push(currentVal);
        currentVal = '';
      } else if (c === '\r') {
        // skip \r
      } else if (c === '\n') {
        currentRow.push(currentVal);
        currentVal = '';
        rows.push(currentRow);
        currentRow = [];
      } else {
        currentVal += c;
      }
    }
  }
  if (currentVal || currentRow.length > 0) {
    currentRow.push(currentVal);
    rows.push(currentRow);
  }
  
  return rows;
}

const allRows = parseCSV(content);
console.log(`Total CSV rows parsed: ${allRows.length}`);

const updateMap = {};

// Data starts at row index 9 (0-based), skipping header rows 0-8
for (let i = 9; i < allRows.length; i++) {
    const cols = allRows[i];
    if (!cols || cols.length <= 17) continue;
    
    const stt = cols[0] ? cols[0].trim() : '';
    if (!stt || isNaN(stt)) continue;
    
    const orderType = cols[1] ? cols[1].trim() : '';
    const productTier = cols[2] ? cols[2].trim() : '';
    const partner = cols[17] ? cols[17].trim() : '';
    
    if (stt && partner) {
        const sttPadded = stt.padStart(4, '0');
        const orderNo = `${sttPadded}/2026/PLMR-${partner}`;
        
        if (!updateMap[orderNo]) {
            updateMap[orderNo] = {
                orderType: orderType,
                productTier: productTier
            };
        }
    }
}

console.log(`Prepared mapping for ${Object.keys(updateMap).length} orders.`);

if (Object.keys(updateMap).length > 0) {
    const payload = JSON.stringify({
        action: "updateClassifications",
        data: {
            updateMap: updateMap
        }
    });
    
    const options = {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(payload)
        }
    };
    
    console.log("Sending request to Google Apps Script...");
    
    // Node.js doesn't natively follow redirects on POST seamlessly with https.request. 
    // We can use fetch if Node > 18.
    if (typeof fetch !== 'undefined') {
        fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: payload,
            redirect: 'follow'
        })
        .then(res => res.json())
        .then(data => console.log('Response:', data))
        .catch(err => console.error('Error:', err));
    } else {
        console.log("Node fetch API not available. Cannot send POST with follow-redirect natively. Please use Node v18+");
    }
}

File: sync_classifications.py
import csv
import json
import urllib.request
import urllib.parse
import ssl

csv_file = 'PP_ KHSX TOTAL 26 - DATA Đặt Hàng 2026.csv'
url = 'https://script.google.com/macros/s/AKfycbyCUwPQbji4QRXS4E7KRQ3PERxlu-IByYSdJCXgeLucxFupJukqLq_0CdXKZpC7okKHsQ/exec'

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

update_map = {}

# Naive CSV parser for quoted fields
def parse_csv_line(text):
    ret = []
    val = ''
    in_quote = False
    for c in text:
        if c == '"':
            in_quote = not in_quote
        elif c == ',' and not in_quote:
            ret.append(val)
            val = ''
        else:
            val += c
    ret.append(val)
    return ret

with open(csv_file, 'r', encoding='utf-8-sig') as f:
    lines = f.readlines()

for line in lines[8:]:
    if not line.strip():
        continue
    cols = parse_csv_line(line.strip())
    if len(cols) > 17:
        stt = cols[0].strip()
        if not stt or not stt.isdigit():
            continue
        
        order_type = cols[1].strip()
        product_tier = cols[2].strip()
        partner = cols[17].strip()
        
        if stt and partner:
            stt_padded = stt.zfill(4)
            order_no = f"{stt_padded}/2026/PLMR-{partner}"
            
            if order_no not in update_map:
                update_map[order_no] = {
                    "orderType": order_type,
                    "productTier": product_tier
                }

print(f"Prepared mapping for {len(update_map)} orders.")

if len(update_map) > 0:
    payload = {
        "action": "updateClassifications",
        "data": {
            "updateMap": update_map
        }
    }
    
    data = json.dumps(payload).encode('utf-8')
    req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'}, method='POST')
    
    try:
        with urllib.request.urlopen(req, context=ctx) as response:
            res_body = response.read().decode('utf-8')
            res_json = json.loads(res_body)
            if res_json.get('success'):
                print(f"Success: {res_json.get('message')}")
            else:
                print(f"Failed: {res_json.get('message')}")
    except Exception as e:
        print(f"Error making request: {e}")
